README.md

## Introduction
This is a NodeJS tool to consume a WSDL file and output a neat, manageable Javascript library. It is not 100% complete, but it gives us full coverage of the services we need at Holiday Extras. If it doesn't do everything you need, fork us and contribute back to the community :)

## Where do I find it?
Either get it from npm:
```
sudo npm install -g wsdl2.js
```
Or checkout the repo:
```
git clone https://github.com/holidayextras/wsdl2.js.git
cd wsdl2.js
npm install
```

## How do I use it?
This will generate a folder called [serviceName] in the current directory ready to be require'd and used:
```
wsdl2.js [serviceName] [/local/path/to/wsdl]
```

## SOAP 1.1
It's assumed by default you're working with a SOAP 1.2 service expecting an `application/soap+xml; charset=utf-8` content-type; you can pass in an option to get the `text/xml; charset=utf-8` that a SOAP 1.1 service will be expecting:

```
wsdl2.js [serviceName] [/local/path/to/wsdl] --soap-version 1.1
```

If the service responds with something like "The server cannot service the request because the media type is unsupported.", you are probably targeting the wrong SOAP version.

## Send empty tags
By default any empty tags will be stripped out of the request as they are probably not needed. Some services may require you to send these even if they are empty. You can change this behaviour by passing in the following option:

```
wsdl2.js [serviceName] [/local/path/to/wsdl] --keep-empty-tags
```

## Requirements for using the generated code
Note: these modules are installed by npm-installing wsdl2.js
```
npm install request xml2json
```

## Generated code structure
```
./[ServiceName]/
|   // This holds one file per defined Element within the WSDL
├── Element
│   ├── SomeDefinedElement.js
|   └── ...
|   // This is the main file which handles requests, JSON->XML->JSON, etc
├── index.js
|   // This is where mock data goes from [myService].Settings.createMock
├── Mocks
│   ├── WsdlOperationName.js
|   └── ...
|   // This library provides strong typing, it's used in each Element/Type
├── Modeler.js
|   // This file defines the top level functionality found within the WSDL
├── ServiceDefinition.js
|   // This holds one file per defined Type within the WSDL
└── Type
    ├── SomeDefinedType.js
    └── ...
```

## Using the generated code
Start by including the generated code:
```javascript
var Service = require("path/to/generated/code");
var EC2 = require("lib/EC2");
```
This is how we create a new request:
```javascript
var someRequest = new Service.[WSDL-Binding-Name].[WSDL-Operation-Name]();
var assignRequest = new EC2.AmazonEC2Port.AssignPrivateIpAddresses();
```
Setting basic properties is trivial
```javascript
someRequest.someSimpleProperty = 1;
```
Most requests consist of several complex types, they are all found within our service object:
```javascript
someRequest.someElementProperty = new Service.Elements.[WSDL-Element-Name]();
someRequest.someTypeProperty = new Service.Types.[WSDL-Type-Name]();
```
Populating Requests/Elements/Types can be done one at a time:
```javascript
someRequest.someNumber = 1;
someRequest.someString = "1";
```
We can also populate directly from a JSON object:
```javascript
var json = { someNumber: 1, someString: "1" };
someRequest = new Service.TestRequest(json);
// someRequest.someNumber == 1
// someRequest.someString == "1"
```
Trying to set a property's value to an invalid type will be discarded:
```javascript
someRequest.PersonElement = null;
someRequest.PersonElement = new SomeRandomObject();
// someRequest.PersonElement == null;
```
If we have an array of objects there's a helper function to save typing:
```javascript
someRequest.PeopleList = new Service.Types.ArrayOfPeople;
someRequest.PeopleList.newChild({ firstname: "Oli", age: 24 });
// is the equivalent of:
someRequest.PeopleList = new Service.Types.ArrayOfPeople;
var newPerson = new Service.Types.Person();
newPerson.firstname = "Oli";
newPerson.age = 24;
someRequest.PeopleList.push(newPerson);
```
Making the request is trivial:
```javascript
someRequest.request(function(err, response) {
  // 'response' is a modeled object, it WILL conform to the WSDL.
  //... w00p!
});
```
Once we have a request and we want to edit it by adding properties not found in the WDSL, we must first extract the data from the response:
```javascript
someRequest.request(function(err, response) {
  response.SomeInvalidProperty = "testing";
  // response.SomeInvalidProperty == null
  var myResponse = response.extract();
  myResponse.SomeInvalidProperty = "testing";
  // myResponse.SomeInvalidProperty == "testing"
});
```

## Runtime Settings and Debugging
```javascript
var Service = require("path/to/generated/code");

// This next statement will change the duration (in milliseconds)
// the library will wait for the service to respond before timing
// out. The default is 15000 (15 seconds)
Service.Settings.timeout = 5000;

// This next statement will enable debugging for ALL soap requests
// It'll print out the request start and end times, how long the
// request took, xml that was sent and recieved as well as the
// json recieved and the output provided by the library
// default: false
Service.Settings.debugSoap = true;

// This next statement allows you to override the default (stdout) logging
// provided by wsdl2.js. The function will recieve a json object.
// default: stdout
Service.Settings.logger = function(data) {
  fs.createWriteStream('./test').write(JSON.stringify(data));
};

// This next statement will enable benchmarking for ALL soap requests
// It prints to stdout the name of each request and its duration in ms
// default: true
Service.Settings.benchmark = true;

// This next statement will store the most recent request of each type to file
// It outputs to [/path/to/generated/code]/Mocks/[request-name]
// default: false
Service.Settings.createMock = true;

// This next statement will use saved mock requests instead of real requests
// default: false
Service.Settings.useMock = true;

// We can debug single SOAP requests by using the .debug() function, which is
// a property of every request and response object
var additionRequest = new Service.MathService.AdditionFunction(json);
additionRequest.debug(); // Watch your console output
additionRequest.request(function(err, response) {
  response.debug(); // Watch your console output
  if (err || !response) {
    return callback(err || "No response?");
  }

  //... w00p!
});
```

## Checking the generated service definition
Generic example of ./[ServiceName]/ServiceDefinition.js:
```javascript
module.exports = {
  "[WSDL-Binding-Name]": {
    "namespace": "http://blahblah.com/webservices/",
    "serviceUrl": "https://www.blahblah.com/path/to/service.blah",
    "[WSDL-Operation-Name]": {
      "soapAction": "blahblahblah",
      "input": "[Top-Level-Request-Element-For-Operation]",
      "output": "[Top-Level-Response-Element-For-Operation]"
    },
    ...
  },
  ...
}
```
Example of Amazon's EC2 service:
```javascript
module.exports = {
  "AmazonEC2Port": {
    "namespace": "http://ec2.amazonaws.com/doc/2013-02-01/",
    "serviceUrl": "https://ec2.amazonaws.com/",
    "ActivateLicense": {
      "soapAction": "ActivateLicense",
      "input": "ElementActivateLicense",
      "output": "ElementActivateLicenseResponse"
    },
    "AssignPrivateIpAddresses": {
      "soapAction": "AssignPrivateIpAddresses",
      "input": "ElementAssignPrivateIpAddresses",
      "output": "ElementAssignPrivateIpAddressesResponse"
    },
    ...
  },
  ...
}
```

## Tests
Run tests using `npm test`*

*: You may need to run `ulimit -n 1000` to up the number of open files you can have open at any one time.

CHANGELOG.md
## Changelog

### v0.4.0
- Send an object containing all request and response information to the logger (@rahulpatel)

### v0.3.0
- Allow changing the default timeout value (@shackpank)
- Allow changing the default log output (@shackpank)

### v0.2.1
- Prevent xml2json double-escaping <>()#&" and ' (@shackpank)

### v0.2.0
- SOAP 1.1 support (@shackpank)
- Send empty tags on request support (@rahulpatel)

### v0.1.0
- Initial release


wsdl2.js
#!/usr/bin/env node
/*******************************************************************
Copyright 2013 Holiday Extras

This file is part of wsdl2.js

wsdl2.js is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

wsdl2.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with wsdl2.js.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************/
var Modeler = require("./lib/Modeler.js");
var parser = require('xml2json');
var fs = require('fs');
var minimist = require('minimist');

var typeMap = { };
var ns = { };
var services = { };
var bindings = { };
var messages = { };
var portTypes = { };
var xmlCache = {
  elements: { },
  complexTypes: { },
  simpleTypes: { }
};
var contentTypeHeaders = {
  '1.1': 'text/xml; charset=utf-8',
  '1.2': 'application/soap+xml; charset=utf-8'
};

var argv = minimist(process.argv.slice(2));
var serviceName = argv._[0];
var pathToWsdl = argv._[1];
var soapVersion = (argv['soap-version'] || '1.2').toString();
var keepEmptyTags = argv['keep-empty-tags'] ? true : false;

if(Object.keys(contentTypeHeaders).indexOf(soapVersion) === -1) {
  process.stdout.write('invalid --soap-version option, expected one of ' + Object.keys(contentTypeHeaders).join(', ') + '\n');
  process.exit(1);
}

if(!(serviceName && pathToWsdl)) {
  process.stdout.write('usage: wsdl2.js [serviceName] [/local/path/to/wsdl]\n');
  process.exit(1);
}

try {
  var xmlWsdlDefinition = fs.readFileSync(pathToWsdl);
} catch(e) {
  process.stdout.write(pathToWsdl + ' could not be opened\n');
  process.exit(1);
}

try {
  fs.mkdirSync(process.cwd()+"/"+serviceName);
} catch(e) { }
try {
  fs.mkdirSync(process.cwd()+"/"+serviceName+"/Type");
} catch(e) { }
try {
  fs.mkdirSync(process.cwd()+"/"+serviceName+"/Element");
} catch(e) { }
try {
  fs.mkdirSync(process.cwd()+"/"+serviceName+"/Mocks");
} catch(e) { }

var classTemplate = fs.readFileSync(__dirname+'/lib/classTemplate.js');
var json = JSON.parse(parser.toJson(xmlWsdlDefinition));
processWSDL(json);

function stripNamespace(t) {
  if (t.match(":")) {
    return t.split(":")[1];
  }
  return t;
};

function processWSDL(json) {
  findNamespaces(json);
  processServices(json);
  processMessages(json);
  processPortTypes(json);
  processBindings(json);
  processTypes(json);

  var serviceDefinition = "module.exports = "+JSON.stringify(services,null,2);
  fs.writeFile(process.cwd()+"/"+serviceName+'/ServiceDefinition.js', serviceDefinition);

  var modelerFile = fs.readFileSync(__dirname+"/lib/Modeler.js");
  fs.writeFile(process.cwd()+"/"+serviceName+'/Modeler.js', modelerFile);

  var indexFile = fs.readFileSync(__dirname+"/lib/serviceProvider.js", 'utf8');
  indexFile = indexFile.replace("###1###", contentTypeHeaders[soapVersion]);
  indexFile = indexFile.replace('###2###', keepEmptyTags);
  fs.writeFile(process.cwd()+"/"+serviceName+'/index.js', indexFile);
};

/*****************************************
<wsdl:definitions xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:tm="http://microsoft.com/wsdl/mime/textMatching/" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:mime="http://schemas.xmlsoap.org/wsdl/mime/" xmlns:tns="http://firemelon.com/webservices/" xmlns:s1="http://microsoft.com/wsdl/types/" xmlns:s="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://schemas.xmlsoap.org/wsdl/soap12/" xmlns:http="http://schemas.xmlsoap.org/wsdl/http/" targetNamespace="http://firemelon.com/webservices/" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/">
*****************************************/
function findNamespaces(json) {
  var mapping = {
    'http://schemas.xmlsoap.org/wsdl/soap/': 'soap',
    'http://schemas.xmlsoap.org/wsdl/': 'wsdl',
    'http://schemas.xmlsoap.org/wsdl/http/': 'http',
    'http://www.w3.org/2001/XMLSchema': 'xsd'
  };
  var wsdlDefinitions = json[Object.keys(json)[0]];

  for (var someAttr in wsdlDefinitions) {
    if (mapping.hasOwnProperty(wsdlDefinitions[someAttr])) {
      ns[mapping[wsdlDefinitions[someAttr]]] = (someAttr.split(":").concat([""]))[1]+":";
    }
  };

  if (!ns.hasOwnProperty('wsdl') || (ns['wsdl']==":")) ns['wsdl'] = "";

  typeMap[ns['xsd']+"string"] = "string";
  typeMap[ns['xsd']+"int"] = "number";
  typeMap[ns['xsd']+"boolean"] = "boolean";
  typeMap[ns['xsd']+"double"] = "number";
  typeMap[ns['xsd']+"decimal"] = "number";
  typeMap[ns['xsd']+"dateTime"] = "date";
  //console.log("Namespaces:", ns);
  //console.log("");
};

/******************************************
<wsdl:service name="Service">
  <wsdl:port name="ServiceSoap" binding="tns:ServiceSoap">
    <soap:address location="https://www.blahblah.com/blah/blah/service.asmx"/>
******************************************/
function processServices(json) {
  var wsdlDefinitions = json[ns['wsdl']+'definitions'];
  var wsdlServices = wsdlDefinitions[ns['wsdl']+'service'];
  if (!(wsdlServices instanceof Array)) wsdlServices = [wsdlServices];
  wsdlServices.forEach(function(someService) {
    var portTags = someService[ns['wsdl']+'port'];
    if (!(portTags instanceof Array)) portTags = [portTags];
    portTags.forEach(function(somePort) {
      if (somePort[ns['soap']+'address']) {
        services[somePort.name] = {
          binding: somePort.binding,
          namespace: wsdlDefinitions["xmlns:"+somePort.binding.split(":")[0]],
          serviceUrl: somePort[ns['soap']+'address'].location
        }
      }
    });
  });
  //console.log("Services:", JSON.stringify(services,null,2));
  //console.log("");
};

/******************************************
<wsdl:message name="SomeMessageName">
  <wsdl:part name="parameters" element="tns:SomeElementName"/>
</wsdl:message>
<wsdl:message name="SomeMessageName">
  <wsdl:part name="parameters" element="tns:SomeElementName"/>
</wsdl:message>
******************************************/
function processMessages(json) {
  var wsdlDefinitions = json[ns['wsdl']+'definitions'];
  var wsdlMessages = wsdlDefinitions[ns['wsdl']+'message'];
  if (!(wsdlMessages instanceof Array)) wsdlMessages = [wsdlMessages];
  wsdlMessages.forEach(function(someMessage) {
    // According to the W3 spec, messages can have multiple part's
    messages[someMessage.name] = [];
    var wsdlParts = someMessage[ns['wsdl']+'part'];
    if (wsdlParts != null) {
      if (!(wsdlParts instanceof Array)) wsdlParts = [wsdlParts];
      wsdlParts.forEach(function(somePart) {
        // According to the W3 spec, parts can have either element or type attributes
        // elements are found here: /wsdl:defintions/wsdl:types/s:schema/s:element
        // types are found here /wsdl:defintions/wsdl:types/s:schema/s:[complex/simple]Type
        if (somePart.element) messages[someMessage.name].push("Element"+stripNamespace(somePart.element));
        if (somePart.type) messages[someMessage.name].push("Type"+stripNamespace(somePart.type));
      });
    }
  });
  //console.log("Messages:", messages);
  //console.log("");
};


/******************************************
<wsdl:portType name="ServiceSoap">
  <wsdl:operation name="SomeOperationName">
    <wsdl:input message="tns:SomeMessageName"/>
    <wsdl:output message="tns:SomeMessageName"/>
******************************************/
function processPortTypes(json) {
  var wsdlDefinitions = json[ns['wsdl']+'definitions'];
  var wsdlPortTypes = wsdlDefinitions[ns['wsdl']+'portType'];
  if (!(wsdlPortTypes instanceof Array)) wsdlPortTypes = [wsdlPortTypes];
  wsdlPortTypes.forEach(function(somePortType) {
    portTypes[somePortType.name] = { };
    var wsdlOperations = somePortType[ns['wsdl']+'operation'];
    if (!(wsdlOperations instanceof Array)) wsdlOperations = [wsdlOperations];
    wsdlOperations.forEach(function(someOperation) {
      portTypes[somePortType.name][someOperation.name] = {
        input: stripNamespace(someOperation[ns['wsdl']+'input']['message']),
        output: stripNamespace(someOperation[ns['wsdl']+'output']['message']),
      };
    });
  });
  //console.log("PortTypes:", portTypes);
  //console.log("");
};


/******************************************
<wsdl:binding name="ServiceSoap" type="tns:ServiceSoap">
  <soap:binding transport="http://schemas.xmlsoap.org/soap/http"/>
  <http:binding verb="GET" />
  <wsdl:operation name="SomeOperationName">
    <soap:operation soapAction="http://blahblah.com/blah/blah" style="document"/>
      <wsdl:input>
        <soap:body use="literal"/>
      </wsdl:input>
      <wsdl:output>
        <soap:body use="literal"/>
      </wsdl:output>
******************************************/
function processBindings(json) {
  var wsdlDefinitions = json[ns['wsdl']+'definitions'];
  var wsdlBindings = wsdlDefinitions[ns['wsdl']+'binding'];
  if (!(wsdlBindings instanceof Array)) wsdlBindings = [wsdlBindings];
  wsdlBindings.forEach(function(someBinding) {
    // First up, find the wsdl service this binding relates to
    var thisService;
    for (someService in services) {
      if (stripNamespace(services[someService].binding) == someBinding.name) {
        thisService = services[someService];

        // Now process each operation
        var wsdlOperations = someBinding[ns['wsdl']+'operation'];
        if (!(wsdlOperations instanceof Array)) wsdlOperations = [wsdlOperations];
        wsdlOperations.forEach(function(someOperation) {
          if (someOperation.hasOwnProperty(ns['soap']+'operation')) {
            var portType = portTypes[stripNamespace(someBinding.type)][someOperation.name];
            bindings[someOperation.name] = {
              soapAction: someOperation[ns['soap']+'operation'].soapAction,
              input: messages[portType.input][0],
              output: messages[portType.output][0]
            };
            // Link the operation back to the service
            thisService[someOperation.name] = bindings[someOperation.name];
          }
        });
      }
    };
  });
  // Remove the placeholders now we've linked bindings to services
  for (var someService in services) {
    delete services[someService].binding;
  }
  //console.log("Bindings:", bindings);
  //console.log("");
};


/******************************************
<wsdl:types>
  <s:schema elementFormDefault="qualified" targetNamespace="http://blahblah.com/blah/">
    <s:import namespace="http://microsoft.com/wsdl/types/" />
    <s:element name="SomeElementName">
      <s:complexType>
        <s:sequence>
          <s:element minOccurs="0" maxOccurs="1" name="AnotherElementName" type="tns:ScriptDetailsResponse"/>
        </s:sequence>
      </s:complexType>
    </s:element>
    <s:complexType name="SomeComplexTypeName">
      <s:sequence>
        <s:element minOccurs="0" maxOccurs="1" name="AnotherElementName" type="s:string"/>
      </s:sequence>
    </s:complexType>
    <s:simpleType name="SomeSimpleTypeName">
      <s:restriction base="s:string">
        <s:pattern value="[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"/>
      </s:restriction>
******************************************/
function processTypes(json) {
  var xsdSchemas = json[ns['wsdl']+'definitions'][ns['wsdl']+'types'][ns['xsd']+'schema'];
  if (!(xsdSchemas instanceof Array)) xsdSchemas = [xsdSchemas];
  xsdSchemas.forEach(function(thisSchema) {
    // TODO handle the import - or not, as people don't seem to use this feature...
    //
    var elements = thisSchema[ns['xsd']+'element'];
    if (elements != null) {
      if (!(elements instanceof Array)) elements = [elements];
      elements.forEach(function(someElement) {
        xmlCache.elements[someElement.name] = someElement;
      });
    }

    var complexTypes = thisSchema[ns['xsd']+'complexType'];
    if (complexTypes != null) {
      if (!(complexTypes instanceof Array)) complexTypes = [complexTypes];
      complexTypes.forEach(function(someComplexType) {
        xmlCache.complexTypes[someComplexType.name] = someComplexType;
      });
    }

    var simpleTypes = thisSchema[ns['xsd']+'simpleType'];
    if (simpleTypes != null) {
      if (!(simpleTypes instanceof Array)) simpleTypes = [simpleTypes];
      simpleTypes.forEach(function(someSimpleType) {
        xmlCache.simpleTypes[someSimpleType.name] = someSimpleType;
      });
    }
  });
  for (var someElementName in xmlCache.elements) {
    var someElement = xmlCache.elements[someElementName];
    var propertyDefinition = propertyModelElement(someElement);
    createClass(someElement.name, "Element", propertyDefinition);
  };
  for (var someComplexTypeName in xmlCache.complexTypes) {
    var someComplexType = xmlCache.complexTypes[someComplexTypeName];
    var propertyDefinition = propertyModelElement(someComplexType);
    createClass(someComplexType.name, "Type", propertyDefinition);
  };
  for (var someSimpleTypeName in xmlCache.simpleTypes) {
    var someSimpleType = xmlCache.simpleTypes[someSimpleTypeName];
    var propertyDefinition = propertyModelElement(someSimpleType);
    createClass(someSimpleType.name, "Type", propertyDefinition);
  };
};

// Convert from XSD to propertyModeler
function propertyModelElement(json) {
  var propertyDefinition = { };
  if (json.hasOwnProperty(ns['xsd']+"complexType")) {
    json = json[ns['xsd']+'complexType'];
  }
  if (json.hasOwnProperty(ns['xsd']+"complexContent")) {
    json = json[ns['xsd']+'complexContent'];
  }
  if (json.hasOwnProperty(ns['xsd']+"extension")) {
    json = json[ns['xsd']+'extension'];
    var parent = stripNamespace(json.base);
    if (xmlCache.elements.hasOwnProperty(parent)) {
      parent = xmlCache.elements[parent];
    } else if (xmlCache.complexTypes.hasOwnProperty(parent)) {
      parent = xmlCache.complexTypes[parent];
    } else if (xmlCache.simpleTypes.hasOwnProperty(parent)) {
      parent = xmlCache.simpleTypes[parent];
    } else {
      console.log("Couldn't find extension by name:", parent);
      parent = null;
    }
    if (parent != null) {
      extendProperty(propertyDefinition, propertyModelElement(parent));
    }
  }
  if (json.hasOwnProperty(ns['xsd']+"sequence")) {
    json = json[ns['xsd']+'sequence'];
  }
  if (json.hasOwnProperty(ns['xsd']+"choice")) {
    json = json[ns['xsd']+'choice'];
  }
  if (json.hasOwnProperty(ns['xsd']+"element")) {
    json = json[ns['xsd']+'element'];
  }
  if (json != null) {
    if (!(json instanceof Array)) json = [json];
    for (var i=0; i<json.length; i++) {
      var wsdlProperty = json[i];
      // Set default type?
      if (!wsdlProperty.hasOwnProperty("type")) wsdlProperty.type = ns['xsd']+"string";
      var newType = "";
      if (typeMap.hasOwnProperty(wsdlProperty.type)) {
        newType = typeMap[wsdlProperty.type];
      } else {
        newType = "Type"+stripNamespace(wsdlProperty.type);
      }

      //
      // Build the PropertyModeler property definition
      //
      propertyDefinition[wsdlProperty.name] = {
        type: newType,
        wsdlDefinition: wsdlProperty,
        mask: Modeler.GET | Modeler.SET,
        required: (parseInt(wsdlProperty.minOccurs)>0)
      };
      //propertyDefinition[wsdlProperty.name].wsdlDefinition.type = newType;

      if ((parseInt(wsdlProperty.maxOccurs)>1 || (wsdlProperty.maxOccurs=="unbounded")) &&
          (!newType.match(/ArrayOf/)) ) {
        propertyDefinition[wsdlProperty.name].mask |= Modeler.ARRAY;
      }
      if (wsdlProperty.nillable) {
        propertyDefinition[wsdlProperty.name].required = false;
      }
      // ...
      };
  } else {
    console.error("Uhhh... Whats this?!", JSON.stringify(json, null, 2));
    propertyDefinition.dummy = { };
  };

  return propertyDefinition;
};


function extendProperty(a, b) {
  for (var property in b) {
    a[property] = b[property];
  }
};

function createClass(className, type, propertyDefinition) {
  // Copy the PropertyModeler Class tempate and set the Classname
  var newClass = ""+classTemplate;
  newClass = newClass.replace(/###1###/g, type+className);
  // Drop the PropertyModeler property definitions into the template
  var txt = JSON.stringify(propertyDefinition, null, 2);
  txt = txt.substring(4, txt.length-2);
  txt = txt.replace(/\n/g, "\n  ");
  txt = txt.replace(/"mask": (.*),/g, function(a, m) {
    var result = "\"mask\": Modeler.GET"
    if (m & Modeler.SET) result += " | Modeler.SET";
    if (m & Modeler.ARRAY) result += " | Modeler.ARRAY";
    return result+",";
  });
  txt = txt.replace(/"(.*)":/g, function(a, m) {
    if (m.match(":")) return "\""+m+"\":";
    return m+":";
  });

  newClass = newClass.replace(/###2###/g, txt);
  // Insert any functions into the template
  newClass = newClass.replace(/###3###/g, "");
  // Write the Class definition to file
  try {
    fs.unlinkSync(serviceName+'/'+className+".js", 10000);
  } catch(e) { }
  fs.writeFile(process.cwd()+'/'+serviceName+'/'+type+"/"+className+".js", newClass, function (err) { });
};










